cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/InternalRouterHandler.java

```diff
@@ -0,0 +1,159 @@
+/*
+ * Copyright © 2023 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.internal.app.runtime.monitor;
+
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import com.google.inject.Inject;
+import io.cdap.cdap.common.BadRequestException;
+import io.cdap.cdap.common.ServiceUnavailableException;
+import io.cdap.cdap.common.conf.Constants;
+import io.cdap.cdap.common.discovery.EndpointStrategy;
+import io.cdap.cdap.common.discovery.RandomEndpointStrategy;
+import io.cdap.cdap.common.discovery.URIScheme;
+import io.cdap.cdap.common.security.HttpsEnabler;
+import io.cdap.http.AbstractHttpHandler;
+import io.cdap.http.BodyConsumer;
+import io.cdap.http.HttpResponder;
+import io.netty.handler.codec.http.HttpRequest;
+import io.netty.handler.codec.http.HttpResponseStatus;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.net.HttpURLConnection;
+import java.net.MalformedURLException;
+import java.net.ProtocolException;
+import java.net.URI;
+import java.net.URL;
+import java.net.UnknownServiceException;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import javax.net.ssl.HttpsURLConnection;
+import javax.ws.rs.DELETE;
+import javax.ws.rs.GET;
+import javax.ws.rs.POST;
+import javax.ws.rs.PUT;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+import org.apache.twill.discovery.Discoverable;
+import org.apache.twill.discovery.DiscoveryServiceClient;
+
+/**
+ * The http handler for routing CDAP service requests from workers running in the
+ * same cluster. This service delegates all auth checks to downstream services.
+ */
+@Path(Constants.Gateway.INTERNAL_API_VERSION_3
+    + "/router")
+public class InternalRouterHandler extends AbstractHttpHandler {
```

Comment:
Is this almost identical to the `RuntimeServiceRoutingHandler` except the path annotation on the class and the request validation? Maybe it's worth to move most of the logic into an abstract base class and have different Path annotation for the sub-class and an abstract method for the request validation?

Reply:
Nice suggestion. Instead of having multiple level of inheritance with Abstract classes, AbstractHttpHandler -> ServiceRouterHandler -> InternalRouter/RuntimeRoutingHandler, I used composition. I extracted the routing logic into a `ServiceRouter` class.

Reply:
A small problem with using ServiceRouter is that this class would be full of one line method. What's the problem to have a common abstract handler for this and the runtime service handler?

Reply:
Besides, the ServiceRouter class itself is not a http handler, yet seeing all the methods that deal with HttpRequest and HttpResponse, which doesn't go well with the existing pattern that only HttpHandler are dealing with http request/response.

Reply:
I don't see any problem with using an Abstract Class at the moment. I read that we should prefer using composition over inheritance to have more flexible code. One problem that I imagine will arise with Abstract Classes is that Java prevents multiple inheritance, so if the RuntimeRoutingHandler wanted some shared behaviour with another class, we would need to make more complicated inheritance hierarchies.

I'll switch to Abstract Class.

Reply:
Switched to using an abstract class. I didn't add a method in the abstract class for validation because I can't figure out a common set of parameters for this method. The runtime service needs program run ID, program type etc. while some other extender may have different requirements. I could have a map of strings passed to the validation method, but then it would get rid of static type checks and auto completion during compilation because all the request parameters will need to be added to a map of strings and re-fetched.